[Frontend] track server_load

daniel-salib opened the pull request:
accurate realtime request concurrency tracking. Added the /load api to retrieve the realtime concurrency count.

benchmark_serving.py with the load tracking:
```
============ Serving Benchmark Result ============
Successful requests:                     20000     
Benchmark duration (s):                  581.50    
Total input tokens:                      4516852   
Total generated tokens:                  3751446   
Request throughput (req/s):              34.39     
Output token throughput (tok/s):         6451.36   
Total Token throughput (tok/s):          14218.98  
---------------Time to First Token----------------
Mean TTFT (ms):                          328943.69 
Median TTFT (ms):                        329329.01 
P99 TTFT (ms):                           557303.45 
-----Time per Output Token (excl. 1st token)------
Mean TPOT (ms):                          32.84     
Median TPOT (ms):                        31.75     
P99 TPOT (ms):                           77.55     
---------------Inter-token Latency----------------
Mean ITL (ms):                           32.21     
Median ITL (ms):                         23.08     
P99 ITL (ms):                            63.13     
==================================================
```

benchmark_serving.py without the load tracking:
```
============ Serving Benchmark Result ============
Successful requests:                     20000     
Benchmark duration (s):                  604.39    
Total input tokens:                      4516852   
Total generated tokens:                  3748047   
Request throughput (req/s):              33.09     
Output token throughput (tok/s):         6201.37   
Total Token throughput (tok/s):          13674.76  
---------------Time to First Token----------------
Mean TTFT (ms):                          346449.51 
Median TTFT (ms):                        341887.55 
P99 TTFT (ms):                           581491.11 
-----Time per Output Token (excl. 1st token)------
Mean TPOT (ms):                          33.11     
Median TPOT (ms):                        32.22     
P99 TPOT (ms):                           78.94     
---------------Inter-token Latency----------------
Mean ITL (ms):                           32.74     
Median ITL (ms):                         23.24     
P99 ITL (ms):                            65.95     
==================================================
```

daniel-salib committed:
track server_load
Signed-off-by: Daniel Salib <danielsalib@meta.com>

russellb commented:
This won't work as-is with multiple frontend processes (API servers), but this also won't be the only metric we have to fix. It's noted as one of the challenges, though we know the expected solution. It's mentioned in the design doc linked from https://github.com/vllm-project/vllm/issues/12705

youngkent commented on vllm/entrypoints/openai/api_server.py:
```diff
@@ -893,6 +900,9 @@ async def init_app_state(
     ) if model_config.runner_type == "transcription" else None
     state.task = model_config.task
 
+    state.concurrent_requests_lock = asyncio.Lock()
+    state.concurrent_requests = 0
```
Since this is just one counter update, I think using regular threading.Lock() might be more efficient than asyncio.Lock, avoiding async context switches.

youngkent commented on vllm/entrypoints/utils.py:
```diff
@@ -57,3 +59,43 @@ async def wrapper(*args, **kwargs):
+                        lambda: asyncio.create_task(combined_background()))
+                return response
+
+            except Exception as exc:
```
We could use a finally block to decrement the counter?

youngkent commented on vllm/entrypoints/utils.py:
```diff
@@ -57,3 +59,43 @@ async def wrapper(*args, **kwargs):
+                request.app.state.concurrent_requests += 1
+
+            try:
+                response = await call_next(request)
```
How will http streaming request be handled in this case?

daniel-salib replied:
from what I can tell from testing, starlette will run the background task only after streaming is complete or when the connection is closed

youngkent commented on vllm/entrypoints/openai/api_server.py:
```diff
@@ -346,6 +347,11 @@ async def health(raw_request: Request) -> Response:
     return Response(status_code=200)
 
 
+@router.get("/load")
```
Can we add a unit test for the /load endpoint?

russellb replied:
Why is this a new API endpoint instead of exposing this as a metric in the existing metrics endpoint?

njhill commented:
+1 on making sure there's no performance impact, we've had some nasty middleware-related performance issues before.

daniel-salib commented:
Thanks for the reviews :) Made another pass taking all the feedback into consideration

youngkent commented on vllm/entrypoints/utils.py:
```diff
@@ -57,3 +59,37 @@ async def wrapper(*args, **kwargs):
+                    # Chain decrement after the existing background task
+                    response.background = BackgroundTask(
+                        lambda: asyncio.create_task(decrement()))
+        if exc:
```
We could just do try/finally without the extra handling of the exception part?

youngkent commented on vllm/entrypoints/utils.py:
```diff
@@ -57,3 +59,37 @@ async def wrapper(*args, **kwargs):
     return wrapper
+
+
+class ConcurrentRequestsMiddleware(BaseHTTPMiddleware):
```
NIT: name it as ServerLoadMiddelware not to be consistent?

youngkent commented on vllm/entrypoints/utils.py:
```diff
@@ -57,3 +59,37 @@ async def wrapper(*args, **kwargs):
+                    response.background = BackgroundTask(decrement)
+                else:
+                    # Chain decrement after the existing background task
+                    response.background = BackgroundTask(
```
This is not chaining the task? Do you mean
  response.backgroud.add_task(decrement)
?

daniel-salib replied:
good catch - yes meant to use add_task

youngkent commented on vllm/entrypoints/openai/api_server.py:
```diff
@@ -273,21 +275,28 @@ def mount_metrics(app: FastAPI):
+    server_load = Gauge("server_load", "Current server load",
+                        registry=registry) if registry \
+                  else Gauge("server_load", "Current server load")
+    server_load.set_function(lambda: app.state.server_load_metrics)
```
server load counter might need to be pulled verify frequently to advice load balancing strategy.
Would pulling /metrics endpoint (generate and parsing all metrics) be inefficient and impacting service performance?
Is there a way to just pull one specific prometheus counter, or we should keep the /load endpoint? cc: @russellb

daniel-salib replied:
makes sense - for now I went back to the /load route to keep the metric polling as lightweight as possible

thanks!

youngkent reviewed (commented):
To address the perf concerns of the middleware, could you paste some benchmark data in the PR summary?

daniel-salib commented:
thanks for the review @youngkent 

 resolved all the comments and added a unit test for the /load route.
 
 Also attached the benchmark results showing the latency comparison before and after the middleware.

youngkent commented on vllm/entrypoints/api_server.py:
```diff
@@ -39,6 +39,12 @@ async def health() -> Response:
     return Response(status_code=200)
 
 
+@app.get("/load")
+async def get_server_load() -> Response:
+    """Health check."""
+    return Response(status_code=200)
```
Is this endpoint not supported for non-openai server? If so, should we return an error by default?

daniel-salib replied:
meant to remove this - from the pattern I see its common to have some endpoints such as /ping and /version that are only defined and implemented in the openai server so I followed the pattern for /load and no longer define it in the non-openai server

youngkent reviewed (commented):
Thanks for updating and adding the latency benchmark data!
It seems latency is regressed by ~1% with middleware enabled, which is a bit higher than expected. Can we understand a bit more if the latency is added to TTFT, TPOT, or just added to the end of a http request due to async background task? (benchmark_serving.py should give some more detailed info)
If it's really stat-sig and impacting TTFT, TPOT, we might need to consider exploring a more efficient implementation without using middleware? e.g. Implement increment/decrement directly in API calls in api_server.

daniel-salib commented:
@youngkent ran benchmark_serving.py twice with and without the middleware

Is it safe to assume the discrepancy between the runs is due to random +/- error between the runs and the middleware's impact is negligible?

youngkent commented:
@daniel-salib Seems the benchmark_serving.py data have wide variance, hard to tell if it's really stat-sig. We might want to increase the sample size, meaning more requests per control and test group. Can we test with 20000 requests per group?

daniel-salib commented:
@youngkent  updated the PR description with the benchmark comparison across 20,000 requests per group

youngkent commented:
Thanks @daniel-salib, the variance is still a bit high, but at least there is no conclusive signal showing middleware regress the performance so far.
@simon-mo Do you want to double check the data and see if this implementation looks good to you? thanks

daniel-salib commented:
@youngkent 

actually, I see I was using the random dataset for the benchmarking, which may be the reason we see the variance. I updated the description with the results after I switched to --dataset-name=sharegpt --dataset-path ShareGPT_V3_unfiltered_cleaned_split.json

robertgshaw2-redhat commented:
What type of GPU is the performance test run on?

robertgshaw2-redhat commented on vllm/entrypoints/openai/api_server.py:
```diff
@@ -894,6 +902,9 @@ async def init_app_state(
     ) if model_config.runner_type == "transcription" else None
     state.task = model_config.task
 
+    state.server_load_metrics_lock = threading.Lock()
```
We do not need locks. This is not threaded code this is asyncio code so only one coroutine can run at a time

robertgshaw2-redhat commented on vllm/entrypoints/utils.py:
```diff
@@ -57,3 +59,31 @@ async def wrapper(*args, **kwargs):
+
+class ServerLoadMiddelware(BaseHTTPMiddleware):
+
+    async def dispatch(self, request, call_next):
```
locks are not needed here

daniel-salib commented:
Thanks for the review @robertgshaw2-redhat 

I ran the performance test on 2 x H100 GPU

simon-mo commented:
The serving throughput degradation is pretty serious, almost a 3% drop. We cannot merge this in as-is due to the performance impact. I added arbitrary middleware support here https://github.com/vllm-project/vllm/blob/a7ea35aa677272d36e01388f26801d673cbb680e/vllm/entrypoints/openai/cli_args.py#L185 if you really need it or alternatively figure out a better way

daniel-salib commented:
@youngkent took a different approach that should be much better performance-wise. Updated the description to include the latest benchmarks

simon-mo commented on vllm/entrypoints/openai/api_server.py:
```diff
@@ -266,6 +267,22 @@ async def validate_json_request(raw_request: Request):
         )
 
 
+def increment_server_load(request: Request):
+    request.app.state.server_load_metrics += 1
+
+
+def decrement_server_load(request: Request):
+    request.app.state.server_load_metrics -= 1
+
+
+async def track_server_load_non_streaming(request: Request):
+    increment_server_load(request)
+    try:
+        yield
+    finally:
+        decrement_server_load(request)
```
the inc and dec can be inlined

simon-mo reviewed (approved):
I think this is a good approach now. Two questions
1. Please put this behind a feature flag so by default there is no perf regression
2. Please define the semantics of "load", which endpoint does it actually cover? I'm confused on why tokenize endpoint is being covered here as it doesn't hit GPU at all.

youngkent commented on vllm/entrypoints/openai/api_server.py:
```diff
@@ -347,13 +364,23 @@ async def health(raw_request: Request) -> Response:
+@router.post("/tokenize",
+             dependencies=[
+                 Depends(validate_json_request),
+                 Depends(track_server_load_non_streaming)
```
Yeah, it might be better to NOT count tokenizer as the load.

youngkent commented on vllm/entrypoints/openai/api_server.py:
```diff
@@ -402,42 +433,66 @@ async def show_version():
         return base(raw_request).create_error_response(
             message="The model does not support Chat Completions API")
 
     generator = await handler.create_chat_completion(request, raw_request)
```
If create_chat_completion throw exception, it would not decrement server load?

daniel-salib commented:
@youngkent  handling exceptions on streaming requests now - PTAL

youngkent commented on vllm/entrypoints/openai/api_server.py:
```diff
@@ -398,46 +454,91 @@ async def show_version():
     generator = await handler.create_chat_completion(request, raw_request)
 
     if isinstance(generator, ErrorResponse):
+        if raw_request.app.state.enable_server_load_tracking:
```
The code is a bit more complex to read now. How about we just make a wrapper function like

```
async def create_chat_completion(...):
    if not raw_request.app.state.enable_server_load_tracking:
        return create_chat_completion(...)
    else: 
        increment_server_load(raw_request)
        try:
            response = create_chat_completion(...)
            response.background = BackgroundTask(...)
            return response
        except:
            decrement_server_load(raw_request)
```

We can further make the above common logic a method like load_aware_call(http_method, ...)

daniel-salib replied:
yea looks much cleaner - I made another pass with "load_aware_streaming_call" implemented. I only applied it to steaming calls because I thought it might be unnecessary for the non streaming requests. The non streaming calls still use the dependency injection.

LMK how it looks and if its better for everything to use load_aware_call

thanks!

youngkent commented on vllm/entrypoints/utils.py:
```diff
@@ -57,3 +59,50 @@ async def wrapper(*args, **kwargs):
+        try:
+            response = await func(*args, raw_request=raw_request, **kwargs)
+            if isinstance(response, (JSONResponse, StreamingResponse)):
+                response.background = BackgroundTask(decrement_server_load,
```
For safety, add assertion that response.background is none?

daniel-salib replied:
I went ahead and handled both cases with and without a pre-existing bg task by adding a task if it already exists

youngkent commented on vllm/entrypoints/utils.py:
```diff
@@ -57,3 +59,50 @@ async def wrapper(*args, **kwargs):
+    request.app.state.server_load_metrics -= 1
+
+
+async def track_server_load_non_streaming(request: Request):
```
Wondering if we could unify the track_server_load_non_streaming and load_aware_streaming_call into a "@load_aware" annotation, put in front of the http methods?

daniel-salib replied:
sounds good - I implemented as a decorator now and removed the dependency injection

daniel-salib commented:
implemented @youngkent 's latest suggestions

also re-ran the benchmark and updated the PR comment to ensure no regression

youngkent commented on vllm/entrypoints/utils.py:
```diff
@@ -57,3 +59,36 @@ async def wrapper(*args, **kwargs):
+                response.background = BackgroundTask(decrement_server_load,
+                                                     raw_request)
+            else:
+                response.background.add_task(decrement_server_load,
```
The BackgroundTask type does not have add_task method? https://github.com/encode/starlette/blob/49158732f8a96744889f5a7644a42ae08131bd15/starlette/background.py#L17

daniel-salib replied:
good catch - I added handling to convert to from a single BackgroundTask -> BackgroundTasks which will us to chain multiple tasks together when needed. Did manual testing locally to confirm its working

youngkent reviewed (approved):
One more comment. Otherwise, looks good. Stamping.

simon-mo commented:
Pre-commit (i.e. lint) is failing.

daniel-salib commented:
@simon-mo  rebased and the precommit is passing now

daniel-salib commented:
@simon-mo I see some failures in the entrypoints test but they seem unrelated to the PR:

I think we're good to merge but LMK if this is an issue. Thanks!

`
[2025-03-11T23:16:51Z] =========================== short test summary info ============================
--
  | [2025-03-11T23:16:51Z] FAILED entrypoints/openai/test_metrics.py::test_metrics_counts[-True] - AssertionError: vllm:iteration_tokens_total_sum expected value of 200 did not match found value 219.0
  | [2025-03-11T23:16:51Z] assert 219.0 == 200
  | [2025-03-11T23:16:51Z]  +  where 219.0 = Sample(name='vllm:iteration_tokens_total_sum', labels={'model_name': 'TinyLlama/TinyLlama-1.1B-Chat-v1.0'}, value=219.0, timestamp=None, exemplar=None).value
  | [2025-03-11T23:16:51Z] FAILED entrypoints/openai/test_metrics.py::test_metrics_counts[-False] - AssertionError: vllm:time_to_first_token_seconds_count expected value of 10 did not match found value 11.0
  | [2025-03-11T23:16:51Z] assert 11.0 == 10
  | [2025-03-11T23:16:51Z]  +  where 11.0 = Sample(name='vllm:time_to_first_token_seconds_count', labels={'model_name': 'TinyLlama/TinyLlama-1.1B-Chat-v1.0'}, value=11.0, timestamp=None, exemplar=None).value
  | [2025-03-11T23:16:51Z] FAILED entrypoints/openai/test_metrics.py::test_metrics_counts[--enable-chunked-prefill-False] - AssertionError: vllm:time_to_first_token_seconds_count expected value of 10 did not match found value 11.0
  | [2025-03-11T23:16:51Z] assert 11.0 == 10
  | [2025-03-11T23:16:51Z]  +  where 11.0 = Sample(name='vllm:time_to_first_token_seconds_count', labels={'model_name': 'TinyLlama/TinyLlama-1.1B-Chat-v1.0'}, value=11.0, timestamp=None, exemplar=None).value
  | [2025-03-11T23:16:51Z] FAILED entrypoints/openai/test_metrics.py::test_metrics_counts[--enable-chunked-prefill-True] - AssertionError: vllm:iteration_tokens_total_sum expected value of 200 did not match found value 219.0
  | [2025-03-11T23:16:51Z] assert 219.0 == 200
  | [2025-03-11T23:16:51Z]  +  where 219.0 = Sample(name='vllm:iteration_tokens_total_sum', labels={'model_name': 'TinyLlama/TinyLlama-1.1B-Chat-v1.0'}, value=219.0, timestamp=None, exemplar=None).value
  | [2025-03-11T23:16:51Z] FAILED entrypoints/openai/test_metrics.py::test_metrics_counts[--disable-frontend-multiprocessing-False] - AssertionError: vllm:time_to_first_token_seconds_count expected value of 10 did not match found value 11.0
  | [2025-03-11T23:16:51Z] assert 11.0 == 10
  | [2025-03-11T23:16:51Z]  +  where 11.0 = Sample(name='vllm:time_to_first_token_seconds_count', labels={'model_name': 'TinyLlama/TinyLlama-1.1B-Chat-v1.0'}, value=11.0, timestamp=None, exemplar=None).value
  | [2025-03-11T23:16:51Z] FAILED entrypoints/openai/test_metrics.py::test_metrics_counts[--disable-frontend-multiprocessing-True] - AssertionError: vllm:iteration_tokens_total_sum expected value of 200 did not match found value 219.0
  | [2025-03-11T23:16:51Z] assert 219.0 == 200
  | [2025-03-11T23:16:51Z]  +  where 219.0 = Sample(name='vllm:iteration_tokens_total_sum', labels={'model_name': 'TinyLlama/TinyLlama-1.1B-Chat-v1.0'}, value=219.0, timestamp=None, exemplar=None).value

`

DarkLight1337 commented:
I think the entrypoints failures are related to this PR, please fix them.

daniel-salib commented:
ah I found the issue, adding the extra unit test to test_metrics messed up the expected metric numbers in metric counts because the tests run on the same server instance. Decided it would be best to keep the metrics test just for prometheus metrics and felt is may be more appropriate to test the server_load arg in test_basic
